sdks/python/apache_beam/runners/worker/sdk_worker.py

```diff
@@ -199,21 +195,26 @@ def task():
         "Currently using %s threads." % len(self._process_thread_pool._threads))
 
   def _request_process_bundle_split(self, request):
-    self._request_process_bundle_action(request)
+    self._request_process_bundle_action(request, self._progress_thread_pool)
+
+  def _request_finalize_bundle(self, request):
+    self._request_process_bundle_action(request, self._finalize_thread_pool)
 
   def _request_process_bundle_progress(self, request):
-    self._request_process_bundle_action(request)
+    self._request_process_bundle_action(request, self._progress_thread_pool)
 
-  def _request_process_bundle_action(self, request):
+  def _request_process_bundle_action(self, request, thread_pool):
 
     def task():
       instruction_reference = getattr(
           request, request.WhichOneof('request')).instruction_reference
-      if instruction_reference in self._instruction_id_vs_worker:
-        self._execute(
-            lambda: self._instruction_id_vs_worker[
-                instruction_reference
-            ].do_instruction(request), request)
+      if instruction_reference not in self._unscheduled_process_bundle:
+        worker = self.workers.get()
```

Comment:
Having available workers here may still be a bottleneck, and the key idea is that we don't want progress/split calls to be blocked on process/finalize ones to complete. 

It should be sufficient to create one extra worker on line 101. Could you drop a TODO there about whether there is value in caching/re-using workers at all?

Reply:
With creating one more worker to handle finalize call sharing the process_thread_pool, it would be possible that the process is blocked by finalize, or the finalize is blocked by process(which seems to be OK).

Reply:
caching/re-using worker would be helpful for progress/split. We can still maintain a worker cache map only for progress/split request.

Reply:
Based on previous discussion: 
> The only issue is that there is a limited number of workers, and we don't want to use them up (or wait for them) for progress/split calls. Given that these worker are stateless and threadsafe, perhaps a reasonable solution would be to create one extra worker that we always use for progress and split (but not for process or bundle finalization).
> 
How about, creating one extra worker inside progress_pool to handle progress/split, creating one extra worker in process_pool to handle finalization. The ```self._worker_count``` will be number of workers processing bundle.

Reply:
The concurrency is handled by two different caps, (1) the number of worker instances created and (2) the size of the threadpools. They don't really live in either pool (despite looking like they do, that's why I suggested a TODO to clean this relationship up). 

If we have number of workers >= size of process pool + size of progress pool, we should be fine. So, yes, let's allocate size of process pool + size of progress pool workers. It's fine for finalize to allocate slot(s) in the progress bundle pool.

Reply:
Implemented here: https://github.com/apache/beam/pull/7937/commits/a67e840c7208047f195de04f1c4deb280a4d97d4